### PR TITLE
fix: proper vitest ts support (#515)

### PR DESCRIPTION
**Pull request by nyaapass**
**What**:

This PR [extends Vitest's default `Assertion` interface](https://vitest.dev/guide/extending-matchers.html#:~:text=If%20you%20are%20using%20TypeScript%2C%20since%20Vitest%200.31.0%20you%20can%20extend%20default%20Assertion%20interface%20in%20an%20ambient%20declaration%20file%20(e.g%3A%20vitest.d.ts)%20with%20the%20code%20below%3A) to fix the `toBeInTheDocument` ts error when using Vitest (#515).

**Why**:

I don't know (still cannot reproduce #515 on my own computer). But [from this comment](https://github.com/testing-library/jest-dom/issues/515#issuecomment-1987823474), adding 

```ts
/// <reference types="@testing-library/jest-dom" />
``` 

in custom `vitest-setup.ts` file would fix this error in that situation. Therefore I think the `types/vitest.d.ts` may need to be updated.

And when I install this repo's dependencies by pnpm and running `tsc -p types/__tests__/vitest`, there would be similar errors of #515 (if using npm or `import '../../jest'` instead of `import '../../vitest'` like that comment, there would be no error). This PR would fix this error even using pnpm.

image

**How**:

https://vitest.dev/guide/extending-matchers.html

**Checklist**:

- [x] Documentation
- [x] Tests
- [x] Updated Type Definitions
- [x] Ready to be merged

### CODE REVIEW

**Comment by ecoms-ye** [comment]
Confirmed in my office computer (a pnpm monorepo with the [With Vitest](https://github.com/testing-library/jest-dom?tab=readme-ov-file#with-vitest) and [With TypeScript](https://github.com/testing-library/jest-dom?tab=readme-ov-file#with-typescript) setup): 

- Switch to the fork version by changing `package.json` to `"@testing-library/jest-dom": "https://github.com/nyaapass/jest-dom"`, the error can be fixed:
    - image
- When back to the original `@testing-library/jest-dom@6.4.2` version, the #515 error would be shown again: 
    - image

**Comment by Dalcvi** [comment]
What is the reason that this change is not being merged in? Vitest removed `@vitest/expect` interface, so when `@testing-library/jest-dom/vitest` is trying to extend `@vitest/expect` interface, it just does nothing. By the way I see it, this change actually corrects the way you are expected to extend the interface since 0.x.x version, so what is the hold up here?

File: types/vitest.d.ts [hunk 1]
```diff
@@ -1,11 +1,20 @@
-import {type expect} from 'vitest'
+/* eslint-disable @typescript-eslint/no-shadow */
+
+import {
+  type Assertion,
+  type AsymmetricMatchersContaining,
```
Comment (Dalcvi) [inline types/vitest.d.ts:5]: These two are unused.

You actually type out the name but don't use it in the line 14 and 19.

**Review by Dalcvi — CHANGES_REQUESTED** [review]

**Comment by Dalcvi** [comment]
> What is the reason that this change is not being merged in? Vitest removed `@vitest/expect` interface, so when `@testing-library/jest-dom/vitest` is trying to extend `@vitest/expect` interface, it just does nothing. By the way I see it, this change actually corrects the way you are expected to extend the interface since 0.x.x version, so what is the hold up here?

Alright, other PR was changed with these changes https://github.com/testing-library/jest-dom/commit/06d19ba759a5ff4b51417cc9a7ecb72c785eb74d waiting for release I guess

**Comment by gnapse** [comment]
I guess we can close this as superseded by #612

**Pull request closed without merging**